semantic/solve.go

```diff
@@ -107,6 +93,89 @@ func (sol *Solution) solve() error {
 	return nil
 }
 
+// unifyKinds will unify the kinds from the kind constraints
+// into a single kind per type variable.
+//
+// The kinds will be unified so that kinds that depend on other
+// type variables will be unified after the kinds for the dependents
+// are unified.
+func (sol *Solution) unifyKinds(kinds map[Tvar]Kind) (Substitution, error) {
+	// Create substitution.
+	subst := make(Substitution)
+
+	// Iterate through each of the kind constraints
+	// and set the ones that only have one constraint
+	// since they do not have to be unified.
+	// Mark down which constraints we have not visited.
+	unvisited := make(map[Tvar]bool, len(sol.cs.kindConst))
+	for tv, ks := range sol.cs.kindConst {
+		if len(ks) == 1 {
+			kinds[tv] = ks[0]
+		} else {
+			unvisited[tv] = true
+		}
+	}
+
+	// Continuously iterate through the kind constraints
+	// and unify any kinds where all dependencies have
+	// already been unified.
+	for len(unvisited) > 0 {
+		// Track if we have visited at least one kind
+		// to avoid a recursive type.
+		once := false
+		for tvl := range unvisited {
+			if !unvisited[tvl] {
+				continue
+			}
+
+			// We may want to visit this constraint.
+			// Check if any other unvisited tvar
+			// occurs in this type.
+			if canVisit := func() bool {
+				for tvr := range unvisited {
+					if tvl == tvr {
+						continue
+					}
+
+					ks := sol.cs.kindConst[tvr]
```

Comment:
Did you mean `ks := sol.cs.kindConst[tvl]`?

Reply:
No. The `unvisited` map contains all of the `Tvar` values that I still need to resolve. So here I am checking that every `Tvar` that hasn't been visited is not contained within the current kind constraints.

Reply:
We discussed in slack that this should be `ks := sol.cs.kindConst[tvl]` and so I agree this algorithm is correct.